src/Utilities/copy.jl

```diff
@@ -292,24 +307,29 @@ function default_copy_to(dest::MOI.ModelLike, src::MOI.ModelLike, copy_names::Bo
     vector_of_variables_types = [S for (F, S) in constraint_types
                                  if F == MOI.VectorOfVariables]
 
-    vector_of_variables_not_added = [
-        copy_vector_of_variables(dest, src, idxmap, S)
-        for S in vector_of_variables_types
-    ]
-    single_variable_not_added = [
-        copy_single_variable(dest, src, idxmap, S)
-        for S in single_variable_types
-    ]
-
-    if length(vis_src) != length(keys(idxmap.varmap))
-        # Copy free variables
-        variables_not_added = setdiff(Set(vis_src), keys(idxmap.varmap))
-        vars = MOI.add_variables(dest, length(variables_not_added))
-        for (vi, var) in zip(variables_not_added, vars)
-            idxmap.varmap[vi] = var
-        end
+    # The `NLPBlock` assumes that the order of variables does not change (#849)
+    if MOI.NLPBlock() in MOI.get(src, MOI.ListOfModelAttributesSet())
+        vector_of_variables_not_added = [
```

Comment:
Why do this only for `NLPBlock` and not always?

Reply:
You mean first determine how each variable is going to be added (either free or constrained) and then add them in the same order (hence splitting the `add_variables` call in separate one for each contiguous block) so that they are added in the same order even if some variables are contrained ?

Note that with https://github.com/JuliaOpt/MathOptInterface.jl/issues/846, the need to not change the order of the variables might go away. The assumption that variables have indices 1 to n that is used by NLP in MOI does not match the rest of the MOI ecosystem and it will go away once `NonlinearExpressionFunction` implements `map_variables`.